exonum-java-binding-common/src/main/java/com/exonum/binding/common/proofs/list/ListProofStructureValidator.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2018 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.common.proofs.list;
+
+import com.exonum.binding.common.proofs.common.ProofStatus;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * A validator that checks list proofs internal structure.
+ */
+public final class ListProofStructureValidator implements ListProofVisitor {
+
+  public static int MAX_NODE_DEPTH = 64;
+
+  private List<NodeInfo> listProofBranchesInfo;
```

Comment:
Please add an issue in "Java Tech Debt" epic to improve its runtime characteristics and a TODO referencing it. This implementation requires O(num_nodes) space and time instead of O(num_elements) space and O(num_nodes) time.

Reply:
Ok, will do.